File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.cc
#include <iostream>
#include <vector>
using namespace std;

int white = 0;
int blue = 0;
vector<vector<int>> board;

void partition(int row, int col, int size) {
    int color = board[row][col];

    for (int i = row; i < row + size; i++) {
        for (int j = col; j < col + size; j++) {
            if (board[i][j] != color) {
                int newSize = size / 2;
                partition(row, col, newSize);
                partition(row, col + newSize, newSize);
                partition(row + newSize, col, newSize);
                partition(row + newSize, col + newSize, newSize);
                return;
            }
        }
    }

    if (color == 0)
        white++;
    else
        blue++;
}

int main() {
    int N;
    cin >> N;

    board = vector<vector<int>>(N, vector<int>(N));

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cin >> board[i][j];
        }
    }

    partition(0, 0, N);

    cout << white << endl;
    cout << blue << endl;

    return 0;
}

File: 백준/Silver/2630. 색종이 만들기/색종이 만들기.py
N = int(input())
board = [list(map(int, input().split())) for _ in range(N)]


def is_paper(row, col, size):
    color = board[row][col]

    for i in range(row, row + size):
        for j in range(col, col + size):
            if board[i][j] != color:
                return False

    return True


blue = 0
white = 0


def dfs(row, col, size):
    global blue, white

    if is_paper(row, col, size):
        if board[row][col] == 0:
            white += 1
        else:
            blue += 1
        return

    dfs(row, col, size // 2)
    dfs(row, col + size // 2, size // 2)
    dfs(row + size // 2, col, size // 2)
    dfs(row + size // 2, col + size // 2, size // 2)


dfs(0, 0, N)
print(white)
print(blue)